File: typescript/packages/core/test/thread/control-item-kit.ts
import { z } from "zod";
import type { Principal } from "../../src/log";
import type { SqlValue, StoreDriver, Tx } from "../../src/store";
import { reading } from "../../src/store/driver";
import { Crash } from "../team/crash-kit";

// The cross-process cancel drills (lane 29F) share these: a second process over the same
// database, the inbox rows to assert on, and a kill at the commit point right after the item's
// insert.

export const OPERATOR: Principal = {
  issuer: "api",
  tenant: "local",
  subject: "operator",
};

export type Row = {
  readonly inbox_id: number;
  readonly channel: string;
  readonly item: Uint8Array;
  readonly consumed_seq: number | null;
};

const Row: z.ZodType<Row> = z.strictObject({
  inbox_id: z.number(),
  channel: z.string(),
  item: z.instanceof(Uint8Array),
  consumed_seq: z.number().nullable(),
});

/** Every inbox row of the store, oldest first. */
export async function inboxRows(db: StoreDriver): Promise<readonly Row[]> {
  return z
    .array(Row)
    .parse(
      await reading(db, (tx: Tx) =>
        tx.all(
          "SELECT inbox_id, channel, item, consumed_seq FROM inbox ORDER BY inbox_id",
          [],
        ),
      ),
    );
}

/** The dead process's leases are gone: what a restart finds. */
export async function expireLeases(db: StoreDriver): Promise<void> {
  await db.transaction((tx) => tx.run("UPDATE leases SET expires_at = 0", []));
}

/**
 * The same database through a driver that dies once on the first statement after an `inbox`
 * insert: that row is committed, and its requester never runs again.
 */
export function crashAfterInsert(base: StoreDriver): StoreDriver {
  let armed = false;
  let crashed = false;
  const wrap = (tx: Tx): Tx => ({
    ...tx,
    run: async (sql: string, params: readonly SqlValue[] = []) => {
      if (armed && !crashed) {
        crashed = true;
        throw new Crash("killed after the inbox item's insert");
      }
      const done = await tx.run(sql, params);
      armed ||= sql.includes("INSERT INTO inbox");
      return done;
    },
    all: async (sql: string, params: readonly SqlValue[] = []) => {
      if (armed && !crashed) {
        crashed = true;
        throw new Crash("killed after the inbox item's insert");
      }
      return tx.all(sql, params);
    },
    transaction: (fn) => tx.transaction((inner) => fn(wrap(inner))),
  });
  return {
    ...base,
    transaction: (fn, options) =>
      base.transaction((tx) => fn(wrap(tx)), options),
  };
}

File: python/tests/thread/control_item_kit.py
"""Shared by the cross-process cancel drills (lane 29F): the inbox rows to assert on, the dead
holder's expired leases, and a runtime the next lease taker builds over an existing branch."""

import sqlite3
from contextlib import closing
from dataclasses import dataclass
from pathlib import Path

from corpus import Clock
from kit import Tools, allow_all

from threads.log import BranchId
from threads.loop.model import Model
from threads.loop.runtime import Runtime, serving
from threads.result import Ok
from threads.store import SqliteStore
from threads.store.sql import int_of, text_of


@dataclass(frozen=True, slots=True)
class Row:
    inbox_id: int
    channel: str
    consumed_seq: int | None


def rows(path: Path) -> tuple[Row, ...]:
    """Every inbox row of the store file, oldest first, read on a connection of its own."""
    with closing(sqlite3.connect(path)) as conn:
        found = conn.execute(
            "SELECT inbox_id, channel, consumed_seq FROM inbox ORDER BY inbox_id"
        ).fetchall()
    return tuple(Row(int_of(i), text_of(c), None if s is None else int_of(s)) for i, c, s in found)


def expire_leases(path: Path) -> None:
    """The dead process's leases are gone: what a restart finds."""
    with closing(sqlite3.connect(path)) as conn, conn:
        conn.execute("UPDATE leases SET expires_at = 0")


async def resumed(
    sq: SqliteStore, branch: BranchId, model: Model, clock: Clock, holder: str = "restart"
) -> Runtime:
    """A runtime over an existing branch, as the next lease taker builds one."""
    taken = await sq.acquire(branch, holder, clock)
    assert isinstance(taken, Ok), taken
    return Runtime(
        sq, taken.value, serving(model), Tools({}, clock), allow_all, clock, clock.wait_until
    )
